src/index/Index.cpp

```diff
@@ -729,6 +736,157 @@ void Index::createPatternsImpl(const string& fileName, const ExtVec& vec,
+template void Index::addPredicates<TsvParser>(const string& filename);
+template void Index::addPredicates<NTriplesParser>(const string& filename);
+
+// _____________________________________________________________________________
+Index::ExtVec Index::computeAddedPredicates(const ExtVec& vec) {
+  if (vec.size() == 0) {
+    LOG(WARN) << "Attempt to write an empty index!" << std::endl;
+  }
+
+  std::unordered_map<Id, size_t> EntityCountMap;
+  std::unordered_map<Id, size_t> textOccMap;
+  std::unordered_set<Id> subjectSet;
+  std::unordered_set<Id> predicateSet;
+  std::unordered_set<Id> objectSet;
+
+  LOG(INFO) << "Counting entities..." << std::endl;
+
+  for (ExtVec::bufreader_type reader(vec); !reader.empty(); ++reader) {
+    ++EntityCountMap[(*reader)[0]];
+    if ((*reader)[1] != (*reader)[0]) {
+      ++EntityCountMap[(*reader)[1]];
+    }
+    if ((*reader)[2] != (*reader)[0] && (*reader)[2] != (*reader)[1]) {
+      ++EntityCountMap[(*reader)[2]];
+    }
+    subjectSet.emplace((*reader)[0]);
+    predicateSet.emplace((*reader)[1]);
+    objectSet.emplace((*reader)[2]);
+  }
+
+  if (_contextFile.size() > 0) {
+    countTextOccurrences(&textOccMap);
+  } else {
+    LOG(INFO) << "No wordsfile submitted. Skipping ql:num-occurrences stat"
+              << std::endl;
+  }
+
+  ExtVec stats(EntityCountMap.size() + subjectSet.size() + predicateSet.size() +
+               objectSet.size() + textOccMap.size());
+  ExtVec::bufwriter_type writer(stats);
+
+  for (auto it = EntityCountMap.begin(); it != EntityCountMap.end(); ++it) {
+    writer << array<Id, 3>{{it->first, Id(0), it->second}};
+  }
+
+  for (auto it = subjectSet.begin(); it != subjectSet.end(); ++it) {
+    writer << array<Id, 3>{{*it, Id(1), Id(0)}};
+  }
+
+  for (auto it = predicateSet.begin(); it != predicateSet.end(); ++it) {
+    writer << array<Id, 3>{{*it, Id(1), Id(1)}};
+  }
+
+  for (auto it = objectSet.begin(); it != objectSet.end(); ++it) {
+    writer << array<Id, 3>{{*it, Id(1), Id(2)}};
+  }
+
+  for (auto it = textOccMap.begin(); it != textOccMap.end(); ++it) {
+    writer << array<Id, 3>{{it->first, Id(2), it->second}};
+  }
+  writer.finish();
+  LOG(INFO) << "Done collecting info for added predicates.\n";
+  return stats;
+}
+
+// _____________________________________________________________________________
+void Index::countTextOccurrences(std::unordered_map<Id, size_t>* textOccMap) {
+  LOG(INFO) << "Making pass over ContextFile " << _contextFile
+            << " and counting text occurrences.\n";
+  ContextFileParser::Line line;
+  ContextFileParser p(_contextFile);
+  size_t i = 0;
+  // write using vector_bufwriter
+
+  // we have deleted the vocabulary during the index creation to save ram, so
+  // now we have to reload it.
+  if (_vocab.size() == 0) {
+    LOG(INFO) << "Loading vocabulary from disk (needed for correct Ids in text "
+                 "index)\n";
+    _vocab = Vocabulary<CompressedString>();
+    readConfigurationFile();
```

Comment:
You may not read the configuration file before it is written!
Currently this is done really at the end of createFromFile AFTER your call to calculateAddedPredicates, that's why the assertion fails. I would like to keep it as late as possible to collect all possible changes we are doing in the future.

Reply:
Thanks for the hint. I put added predicate computation to the end of createFromFile, right after the config file is written